### ISSUE TITLE
Add helper for argument validation

### ISSUE THREAD

**Issue by pascalberger** [cake-build/cake#1561]
Each Cake addin requires some kind of argument validation (or at least should :). Instead of having to implement this in every addin it would be nice if this could be provided for common cases (null check, empty string check, ...) by Cake.Core.

**Comment by pascalberger** [comment]
This could also be helpful Cake internally to simplify argument checking, like its eg done [here](https://github.com/cake-build/cake/blob/develop/src/Cake.Core/CakeContext.cs#L38).

If this is of interest I can provide a PR for a class similar to the one which we use in the [Cake.Prca Addin](https://github.com/cake-contrib/Cake.Prca/blob/develop/src/Cake.Prca/PrcaArgumentChecks.cs).

**Comment by gep13** [comment]
@pascalberger I would have no objection to this.

@cake-build/cake-team thoughts?

**Comment by phillipsj** [comment]
This sounds pretty awesome if it make it easier.  It is a common pattern in every addin.

**Comment by daveaglick** [comment]
I agree, this would be great. There's already several argument validation frameworks - I wonder if any of them have source-only NuGet packages...

**Comment by agc93** [comment]
Agreed that it would be good to centralise this (few other candidates for that) 👍 

@daveaglick I've used Ensure.That from @danielwertheim before, but looks like the source-only version hasn't gotten the latest updates yet..

**Comment by mholo65** [comment]
Agreed! 

But instead of signature ```void Check<T>(T...)``` I would rather see ```T Check<T>(T...)```. Then you could write checks like
```
_foo = Check.NotNull(foo);
```

I personally think ASP.NET team have a quite decent boilerplate [here](https://github.com/aspnet/EntityFramework/blob/dev/src/Shared/Check.cs) which I have used from time to time.

**Comment by agc93** [comment]
Additional reason to centralise this I just noticed: the `Cake.Prca` validation examples @pascalberger linked to currently show up on just about every type in the documentation, which could be a bit misleading..

**Comment by patriksvensson** [comment]
I like the link @mholo65 posted.

**Comment by pascalberger** [comment]
OK, I'll create a PR containing a class based on the ASP.NET boilerplate.

**Comment by patriksvensson** [comment]
@pascalberger Just make sure to attribute ASP.NET project in your code.

**Comment by pascalberger** [comment]
Since there's currently [only interest for aliases providing argument checking](https://github.com/cake-build/cake/pull/1586#issuecomment-298260600) but not also for a public class which can be consumed in addins, I'll leave this to someone else for implementing (or there is also already an implementation for such aliases in [Cake.Incubator](https://github.com/cake-contrib/Cake.Incubator/blob/develop/src/Cake.Incubator/AssertExtensions.cs)).

**Comment by mholo65** [comment]
Adding *Up-for-grabs* on this one.

This is what we thought of:
* Add `internal static` class for argument checking. You could use [this](https://github.com/aspnet/EntityFramework/blob/dev/src/Shared/Check.cs) and attribute ASP.NET team in the comments. This file should be sourced in every project (`<Compile Include="..\Shared\Check.cs" />`)
* Change argument checking to use new approach. E.g. change this:
```
if (foo == null)
{
    throw new ArgumentNullException(nameof(foo));
}
_foo = foo;
```
to:
```
_foo = Check.NotNull(nameof(foo));
```

**Comment by patriksvensson** [comment]
@mholo65 Perhaps we should reserve this one for [https://yourfirstpr.github.io/](https://yourfirstpr.github.io/)? I could do the mentoring.

**Comment by mholo65** [comment]
@patriksvensson sounds good to me!

**Comment by bookman220** [comment]
@patriksvensson I'd be interested in trying this out

**Comment by patriksvensson** [comment]
@bookman220 Great! Do you want any help to get started or do you want to give it a shot yourself?

**Comment by bookman220** [comment]
@patriksvensson I'll probably need some help but let me figure out how to build this thing first.

**Comment by sarithsutha** [comment]
@patriksvensson @bookman220 I don't want to step on anyone's toes, I'm interested in trying this out if it is still up for grabs. This would be my first time contributing to open source.

**Comment by bookman220** [comment]
@sarithsutha @patriksvensson I do apologize for taking so long. I do think I may need some guidance however. Do I just open up the solution file in visual studio and compile it from there? If I haven't fixed the big within the next week you may take it sarithsutha

**Comment by sarithsutha** [comment]
@bookman220 thanks for your response. No rush, take your time, I don't mind waiting. I use Visual Studio 2017, I was able to just open the solution and compile it from there. And with VS code I was able to build using the Cake addin.

**Comment by bookman220** [comment]
@sarithsutha It appears I am having some trouble building the project. Apologies for needing such basic advice but I am new to Visual Studio and C#.  It appears a fair amount of the solution failed to load and when I go to Build it just has the option to run a code analysis on the solution, which I did and you can see the output. Can you give me some pointers as to where to go from here?
![image](https://user-images.githubusercontent.com/5388655/29489635-45907240-84eb-11e7-8565-3b55f4e55ad5.png)

**Comment by patriksvensson** [comment]
@bookman220 There seems to be problems loading the different projects according to the screenshot.

**The following prerequisites need to be installed:**

* Visual Studio 2017 (https://www.visualstudio.com/vs/)
* .NET Core SDK 1.0.4 or later (https://www.microsoft.com/net/download/core)

The `build.cake` file is only used to build the project from command line. When the project is loading correctly in Visual Studio, you should be able to build everything from the `Build` menu.

**Comment by micheleissa** [comment]
@patriksvensson I'd like to take a try on this one.

**Comment by micheleissa** [comment]
What I'm planning to do : 
- Add folder called 'Shared' in the solution directory.
- Add 'Check.cs' inside that folder.
- Modify each csproj file to include 
```
<ItemGroup>
    <Compile Include="..\Shared\Check.cs" />
</ItemGroup>
```
- The namespace for it  `Cake.Shared`
- I saw the link from ASP.NET team but I feel it is more complicated than it should be and I was thinking of something along the following
```
public static T NotNull<T>(T o, string message = null)
            {
            if (o == null)
                {
                throw new ArgumentNullException(message ?? $"{nameof(o)} cannot be null");
                }
            return o;
            }
```
Thoughts?

**Comment by kcamp** [comment]
As a developer, it would be preferable to more correctly support parameter names; `nameof(o)` is not going to be semantically useful in the exception that's thrown.  

Reference:
https://msdn.microsoft.com/en-us/library/k8a0dfcy(v=vs.110).aspx

Consider something like this, or overloads that would support the same.. 
```
T NotNull<T>(T item, string parameterName, string message = null)  { .... }
```

So that the calling convention would be more akin to 
```
public SomeConstructor(object arg) 
{
    _fieldName = ValidationHelperClassName.NotNull(arg, nameof(arg));
    ....
```

**Comment by patriksvensson** [comment]
@micheleissa This is still assigned to @bookman220.
@bookman220 Are you still working on this?

**Comment by micheleissa** [comment]
@kcamp 
What would you use `parameterName` for ?
in the case someone passes 'bogus' as the parameterName and it is used in the exception message the result will not be accurate in my opinion.

**Comment by patriksvensson** [comment]
@micheleissa What @kcamp means is that `nameof(o)` will always print `o`. It will never used the original parameter name.

**Comment by micheleissa** [comment]
@patriksvensson gotcha. Silly me!

**Comment by micheleissa** [comment]
So are you supposed to wait for a response from 'bookman220'

**Comment by patriksvensson** [comment]
@micheleissa let’s give him/her a couple of days to respond before reassigning it.

**Comment by micheleissa** [comment]
sounds good.

**Comment by bookman220** [comment]
@patriksvensson @micheleissa Again I apologize for being so obtuse, but to be honest I am a little disheartened by how flummoxed I am by the environment, which is why it has taken me so long. I work with Javascript scripts in my day job and have debugged compilation errors in C# to get a game running using the Unity game engine, but for some reason I'm finding myself staring slack jawed at this even though it should be simple. At anry rate when I try to build it just says run code analysis on solution, and then in the output below it's all zeros:
![image](https://user-images.githubusercontent.com/5388655/32153147-a7815256-bcf6-11e7-9d04-bb74c3380f68.png)

**Comment by bookman220** [comment]
By the way I downloaded the .NET 2.0 SDK and I already had Visual Studio 2017, though it updated me to the latest version when I ran the installer.

**Comment by bookman220** [comment]
As you can see from my other solution for the games, I'm having no trouble there:
![image](https://user-images.githubusercontent.com/5388655/32153349-f16f7356-bcf7-11e7-90c3-3d7c2bcb6681.png)

**Comment by bookman220** [comment]
Apologies for forgetting to crop out the second screen by the way.

**Comment by kcamp** [comment]
@bookman220 What happens when you run the build [via Cake] from Powershell?
```
c:\git\cake > .\build.ps1
```

**Comment by bookman220** [comment]
Here's the error I get:
![image](https://user-images.githubusercontent.com/5388655/32259648-e3be28da-be8f-11e7-831e-558490b50679.png)

**Comment by bookman220** [comment]
@kcamp @patriksvensson Do you guys have any suggestions?

**Comment by patriksvensson** [comment]
@bookman220 You seem to have a custom nuget source pointing at `myget-2e16-kxcdn.com`. Have you removed the default nuget source?

**Comment by bookman220** [comment]
@patriksvensson I would have done no such thing intentionally. Any tips on how I might restore the default one?

**Comment by jnm2** [comment]
Maybe find the config file that is overriding the defaults? https://docs.microsoft.com/en-us/nuget/consume-packages/configuring-nuget-behavior#config-file-locations-and-uses

**Comment by bookman220** [comment]
@qpooqp Did you tackle this one?

**Comment by qpooqp** [comment]
To be honest i forgot about this one. 
But i will try to do it ASAP.

**Comment by bookman220** [comment]
@patriksvensson I have installed the NuGet Package Manager and am trying to see if "Sync-Package" would fix the issue. Do you know what value I should supply for the "Id" parameter here? 
![image](https://user-images.githubusercontent.com/5388655/48975490-d65b2d80-f036-11e8-84a3-af0ed1f980e0.png)

**Comment by patriksvensson** [comment]
@bookman220 I’m afraod that I have no clue what sync-packages does.

If you’re trying to restore packages for the solution it should be enough to do a rebuild.

**Comment by bookman220** [comment]
@patriksvensson When I go to "Build" I only see "Run Code Analysis on Solution" and when I select that I get "========== Rebuild All: 0 succeeded, 0 failed, 0 skipped =========="
![image](https://user-images.githubusercontent.com/5388655/49323903-be792300-f4e8-11e8-9676-1f7330543df3.png)
![image](https://user-images.githubusercontent.com/5388655/49323909-d355b680-f4e8-11e8-985f-7915faa6dfe5.png)

**Comment by patriksvensson** [comment]
@bookman220 I have no idea why that occurs to you.

Just to double check, do you open `./src/Cake.sln` or did you open something else?
What version of Visual Studio are you using? The latest VS2017 version is recommended.

**Comment by bookman220** [comment]
@patriksvensson 
1) Yes, I'm opening ./src/Cake.sln:
![image](https://user-images.githubusercontent.com/5388655/49337601-6c64fa00-f5db-11e8-8f1f-3be205aae231.png)
2) I had been running a slightly older version (14 something) but now I'm all caught up and have the same issue:
![image](https://user-images.githubusercontent.com/5388655/49337725-1c873280-f5dd-11e8-8330-268697d458f1.png)

**Comment by devlead** [comment]
You'll need the .NET Core sdk bits too.

**Comment by bookman220** [comment]
@devlead Ok now I have the 4.7.2 .NET SDK (added the targeting pack for good measure): 
![image](https://user-images.githubusercontent.com/5388655/49621446-fa9a0100-f98a-11e8-9113-e5688276bde4.png)
One time when I opened it an alert box popped up saying "One or more projects in this solution did not load correctly. See Output for more information", however I cannot reproduce this behavior. Here's what I'm seeing when I open the solution: 
![image](https://user-images.githubusercontent.com/5388655/49621512-4ea4e580-f98b-11e8-9a42-901d2c041522.png)

**Comment by bookman220** [comment]
@patriksvensson I happened to be reading an article on Docker containers and I wondered if that could help people with creating an environment in which myself and others could build the app: https://www.sumologic.com/blog/devops/kubernetes-vs-docker/

**Comment by OhYash** [comment]
Hey is this one done?
@pascalberger posted [AssertExtensions.cs](https://github.com/cake-contrib/Cake.Incubator/blob/develop/src/Cake.Incubator/AssertExtensions.cs) link, which seems to contain the said assertion functions. Still asking cause the issue is open at the moment.
